boot/seal.go

```diff
@@ -883,3 +895,50 @@ func isResealNeeded(pbc predictableBootChains, bootChainsFile string, expectRese
 	}
 	return true, c + 1, nil
 }
+
+func postFactoryResetCleanupSecboot() error {
+	// we are inspecting a key which was generated during factory reset, in
+	// the simplest case the sealed key generated previously used the main
+	// handles, while the current key uses alt handles, hence we need to
+	// release the main handles corresponding to the old key
+	handles := []uint32{secboot.RunObjectPCRPolicyCounterHandle, secboot.FallbackObjectPCRPolicyCounterHandle}
+	usesAlt, err := usesAltPCRHandles()
+	if err != nil {
+		return fmt.Errorf("cannot inspect fallback key: %v", err)
+	}
+	if !usesAlt {
+		// current fallback key using the main handles, which is
+		// possible of there were subsequent factory reset steps,
+		// release the alt handles associated with the old key
+		handles = []uint32{secboot.AltRunObjectPCRPolicyCounterHandle, secboot.AltFallbackObjectPCRPolicyCounterHandle}
+	}
+	return secbootReleasePCRResourceHandles(handles...)
+}
+
+func postFactoryResetCleanup() error {
+	hasHook, err := HasFDESetupHook()
+	if err != nil {
+		return fmt.Errorf("cannot check for fde-setup hook %v", err)
+	}
+
+	saveFallbackKeyFactory := FactoryResetFallbackSaveSealedKeyUnder(InitramfsSeedEncryptionKeyDir)
+	saveFallbackKey := FallbackSaveSealedKeyUnder(InitramfsSeedEncryptionKeyDir)
+	if err := os.Rename(saveFallbackKeyFactory, saveFallbackKey); err != nil {
+		// it is possible that the key file was already renamed if we
+		// came back here after an unexpected reboot
```

Comment:
Should we actually check that `saveFallbackKey` exist, instead of assuming that it does? Or, given that we are cleaning up, it does not matter whether they key is there?